src/serializer/ResidualHeapSerializer.js

```diff
@@ -316,6 +316,8 @@ export class ResidualHeapSerializer {
       invariant(descValue instanceof Value);
       invariant(!this.emitter.getReasonToWaitForDependencies([descValue, val]), "precondition of _emitProperty");
       let mightHaveBeenDeleted = descValue.mightHaveBeenDeleted();
+      // The only case we do not need to remove the dummy property is array index property.
+      const cleanupDummyProperty = !IsArray(this.realm, val) || !IsArrayIndex(this.realm, key);
```

Comment:
This ties the implementation of `_emitProperty` very closely to `_serializeArrayIndexProperties`. Brittle. I'd prefer if there was a new (optional) set being maintained, like `dummyProperties`, which entries are added to by `_serializeArrayIndexProperties`, and `_emitProperty` gets a new parameter `cleanupDummyProperty` which is set by `_emitObjectProperties` as needed.